### PR TITLE
JAMES-3051 Updating Cassandra schema version

### PR DESCRIPTION
**Pull request by tnstung**

### CODE REVIEW

File: src/site/markdown/server/manage-guice-distributed-james.md [hunk 1]
```diff
@@ -493,3 +495,20 @@ Managing an email queue is an easy task if you follow this procedure:
 In case, you need to clear an email queue because there are only spam or trash emails in the email queue you have this procedure to follow:
 
 - All mails from the given mail queue will be deleted with [Clearing a mail queue](manage-webadmin.html#Clearing_a_mail_queue).
+## Updating Cassandra schema version
+
+A schema version indicates you which schema your James server is relying on. The schema version number tracks if a migration is required. For instance, when the latest schema version is 2, and the current schema version is 1, we might think that we still have data in the deprecated Message table in the database. Hence, we need to migrate these messages into the MessageV2 table. Once done, we can safely bump the current schema version to 2.
+
```
Comment (chibenwa) [inline src/site/markdown/server/manage-guice-distributed-james.md]: IMO we need to state that:

Relying on outdated schema version prevents you to benefits from the newest performance and safety  improvments.

File: src/site/markdown/server/manage-guice-distributed-james.md [hunk 2]
```diff
@@ -493,3 +495,20 @@ Managing an email queue is an easy task if you follow this procedure:
...
+### How to detect when we should update Cassandra schema version
+
+When we see in James logs `org.apache.james.modules.mailbox.CassandraSchemaVersionStartUpCheck` shows a warning about Recommend version of Cassandra schema, we should perform to update Cassandra schema version procedures
+
```
Comment (chibenwa) [inline src/site/markdown/server/manage-guice-distributed-james.md]: Also, the Apache PMC keeps track of changes needed when upgrading to a newer version: https://github.com/apache/james-project/blob/master/upgrade-instructions.md

File: src/site/markdown/server/manage-guice-distributed-james.md [hunk 3]
```diff
@@ -24,7 +24,9 @@ advanced users.
  - [Setting Cassandra user permissions](#Setting_Cassandra_user_permissions)
  - [Cassandra table level configuration](#Cassandra_table_level_configuration) 
  - [Mail Queue](#Mail_Queue)
-
+ - [Cassandra table level configuration](#Cassandra_table_level_configuration)
```
Comment (Arsnael) [inline src/site/markdown/server/manage-guice-distributed-james.md]: Careful, mistake in rebase, that entry is also two lines above

File: src/site/markdown/server/manage-guice-distributed-james.md [hunk 4]
```diff
@@ -493,3 +495,20 @@ Managing an email queue is an easy task if you follow this procedure:
 In case, you need to clear an email queue because there are only spam or trash emails in the email queue you have this procedure to follow:
 
 - All mails from the given mail queue will be deleted with [Clearing a mail queue](manage-webadmin.html#Clearing_a_mail_queue).
```
Comment (Arsnael) [inline src/site/markdown/server/manage-guice-distributed-james.md]: Miss line break

File: src/site/markdown/server/manage-guice-distributed-james.md [hunk 5]
```diff
@@ -493,3 +495,20 @@ Managing an email queue is an easy task if you follow this procedure:
...
+## Updating Cassandra schema version
+
+A schema version indicates you which schema your James server is relying on. The schema version number tracks if a migration is required. For instance, when the latest schema version is 2, and the current schema version is 1, we might think that we still have data in the deprecated Message table in the database. Hence, we need to migrate these messages into the MessageV2 table. Once done, we can safely bump the current schema version to 2.
+
+### How to detect when we should update Cassandra schema version
+
+When we see in James logs `org.apache.james.modules.mailbox.CassandraSchemaVersionStartUpCheck` shows a warning about Recommend version of Cassandra schema, we should perform to update Cassandra schema version procedures
```
Comment (Arsnael) [inline src/site/markdown/server/manage-guice-distributed-james.md]: s/shows/showing
s/Recommend/Recommended?

I would maybe reformulate the last part of the sentence like that: `we should perform an update of the Cassandra schema version.`

File: src/site/markdown/server/manage-guice-distributed-james.md [hunk 6]
```diff
@@ -493,3 +495,20 @@ Managing an email queue is an easy task if you follow this procedure:
...
+
+### How to update Cassandra schema version
+
+These schema updates can be triggered by webadmin using the Cassandra backend. Followings are steps to update Cassandra schema:
```
Comment (Arsnael) [inline src/site/markdown/server/manage-guice-distributed-james.md]: `Following steps are for updating the Cassandra schema version:`

File: src/site/markdown/server/manage-guice-distributed-james.md [hunk 7]
```diff
@@ -493,3 +495,20 @@ Managing an email queue is an easy task if you follow this procedure:
...
+
+- At the very first step, we need to [retrieve current Cassandra schema version](manage-webadmin.html#Retrieving_current_Cassandra_schema_version)
+- And then, we [retrieve latest available Cassandra schema version](manage-webadmin.html#Retrieving_latest_available_Cassandra_schema_version) to make sure there is a latest available version
```
Comment (Arsnael) [inline src/site/markdown/server/manage-guice-distributed-james.md]: we `can`

File: src/site/markdown/server/manage-guice-distributed-james.md (hunk 1 above)
Comment (tnstung) [reply src/site/markdown/server/manage-guice-distributed-james.md]: Thanks for your comment. I'll add that statement in the doc.

File: src/site/markdown/server/manage-guice-distributed-james.md (hunk 2 above)
Comment (tnstung) [reply src/site/markdown/server/manage-guice-distributed-james.md]: Thanks. I've added it.

File: src/site/markdown/server/manage-guice-distributed-james.md [hunk 8]
```diff
@@ -493,3 +495,24 @@ Managing an email queue is an easy task if you follow this procedure:
 In case, you need to clear an email queue because there are only spam or trash emails in the email queue you have this procedure to follow:
 
 - All mails from the given mail queue will be deleted with [Clearing a mail queue](manage-webadmin.html#Clearing_a_mail_queue).
+## Updating Cassandra schema version
+
+A schema version indicates you which schema your James server is relying on. The schema version number tracks if a migration is required. For instance, when the latest schema version is 2, and the current schema version is 1, we might think that we still have data in the deprecated Message table in the database. Hence, we need to migrate these messages into the MessageV2 table. Once done, we can safely bump the current schema version to 2.
+
+Relying on outdated schema version prevents you to benefits from the newest performance and safety improvements.
```
Comment (Arsnael) [inline src/site/markdown/server/manage-guice-distributed-james.md]: s/benefits/benefit

File: src/site/markdown/server/manage-guice-distributed-james.md (hunk 5 above)
Comment (tnstung) [reply src/site/markdown/server/manage-guice-distributed-james.md]: You're right:
```
        String upgradeVersionMessage =
            String.format("Current schema version is %d. Recommended version is %d",
                versionManager.computeVersion().getValue(),
                versionManager.getMaximumSupportedVersion().getValue());
        LOGGER.warn(upgradeVersionMessage);
```
Thanks

File: src/site/markdown/server/manage-guice-distributed-james.md [hunk 9]
```diff
@@ -493,3 +494,25 @@ Managing an email queue is an easy task if you follow this procedure:
...
 
 - All mails from the given mail queue will be deleted with [Clearing a mail queue](manage-webadmin.html#Clearing_a_mail_queue).
+
+## Updating Cassandra schema version
+
+A schema version indicates you which schema your James server is relying on. The schema version number tracks if a migration is required. For instance, when the latest schema version is 2, and the current schema version is 1, we might think that we still have data in the deprecated Message table in the database. Hence, we need to migrate these messages into the MessageV2 table. Once done, we can safely bump the current schema version to 2.
+
+Relying on outdated schema version prevents you to benefit from the newest performance and safety improvements.
+
+### How to detect when we should update Cassandra schema version
```
Comment (rouazana) [inline src/site/markdown/server/manage-guice-distributed-james.md]: s/we/you? usually we put we as developpers and you as users/ops. Maybe we can change this bad habit, but then it requires more changes.

File: src/site/markdown/server/manage-guice-distributed-james.md [hunk 10]
```diff
@@ -493,3 +494,25 @@ Managing an email queue is an easy task if you follow this procedure:
...
+
+When we see in James logs `org.apache.james.modules.mailbox.CassandraSchemaVersionStartUpCheck` showing a warning like `Recommended version is versionX`, we should perform an update of the Cassandra schema version.
```
Comment (rouazana) [inline src/site/markdown/server/manage-guice-distributed-james.md]: idem

Comment (rouazana) [reply src/site/markdown/server/manage-guice-distributed-james.md]: (and several times if the next paragraphs)

File: src/site/markdown/server/manage-guice-distributed-james.md [hunk 11]
```diff
@@ -493,3 +494,25 @@ Managing an email queue is an easy task if you follow this procedure:
 In case, you need to clear an email queue because there are only spam or trash emails in the email queue you have this procedure to follow:
 
 - All mails from the given mail queue will be deleted with [Clearing a mail queue](manage-webadmin.html#Clearing_a_mail_queue).
+
+## Updating Cassandra schema version
```
Comment (mbaechler) [inline src/site/markdown/server/manage-guice-distributed-james.md]: There's something very unexpected in the way we manage cassandra schema: we create new tables without asking the admin about it. That means your James version is always using the last tables but may also take into account old ones if migration is not done yet.

It should be carefully described here.

File: src/site/markdown/server/manage-guice-distributed-james.md [hunk 12]
```diff
@@ -493,3 +494,25 @@ Managing an email queue is an easy task if you follow this procedure:
...
+
+A schema version indicates you which schema your James server is relying on. The schema version number tracks if a migration is required. For instance, when the latest schema version is 2, and the current schema version is 1, we might think that we still have data in the deprecated Message table in the database. Hence, we need to migrate these messages into the MessageV2 table. Once done, we can safely bump the current schema version to 2.
+
+Relying on outdated schema version prevents you to benefit from the newest performance and safety improvements.
+
+### How to detect when we should update Cassandra schema version
+
+When we see in James logs `org.apache.james.modules.mailbox.CassandraSchemaVersionStartUpCheck` showing a warning like `Recommended version is versionX`, we should perform an update of the Cassandra schema version.
+
+Also, the Apache PMC keeps track of changes needed when upgrading to a newer version. You can read this [upgrade instructions](https://github.com/apache/james-project/blob/master/upgrade-instructions.md).
```
Comment (mbaechler) [inline src/site/markdown/server/manage-guice-distributed-james.md]: What do you mean by "the Apache PMC"? Why do I need to know that?

Comment (chibenwa) [reply src/site/markdown/server/manage-guice-distributed-james.md]: `Also, we keep track`

Comment (tnstung) [reply src/site/markdown/server/manage-guice-distributed-james.md]: OK thanks :)

File: src/site/markdown/server/manage-guice-distributed-james.md (hunk 9 above)
Comment (tnstung) [reply src/site/markdown/server/manage-guice-distributed-james.md]: I think we should keep you as the above docs used you.

**Comment by Arsnael** [comment]
Merged

**Pull request closed without merging**